### PR TITLE
TGUI: Brig Timer

### PR DESCRIPTION
**Pull request by Crowbar764**
# About the pull request

Upgrades the brig cell timers to use TGUI.
Also fixes the pardon mechanic which was broken by the rank / paygrade overhaul.

Brig timers can now be:

- Started / Resumed
- Paused
- Stopped (Sets timer to the end)
- Reset (Resets timer to 0, removes pardoned status if set)
- Pardoned

COs, High Command, and PvI can pardon.

https://user-images.githubusercontent.com/59719612/209559615-219b243f-cdac-4f14-a572-8ced17460257.mp4

# Changelog

:cl: Casper
fix: fixes pardon mechanic
refactor: brig timer now uses TGUI
/:cl:

### CODE REVIEW

File: code/game/machinery/doors/brig_system.dm [hunk 1]
```diff
@@ -11,348 +17,333 @@
 
 	var/id = null     		// id of door it controls.
 	var/picture_state		// icon_state of alert picture, if not displaying text/numbers
-	var/list/obj/structure/machinery/targets = list()
+	var/list/obj/structure/machinery/targets = list() // A list of World objects this machine controls.
 	var/list/obj/item/paper/incident/incident_reports = list()
-	var/obj/item/paper/incident/current_report = null
-
-	var/current_menu = "main_menu"
+	var/obj/item/paper/incident/active_report = null // The report currently ticking down the clock.
```
Comment (harryob) [inline code/game/machinery/doors/brig_system.dm:22]: i’m aware it’s inconsistent w the rest of the vars here, but if you’re documenting them you might as well dmdoc them so it shows up in vsc hover
```suggestion
/// The report currently ticking down the clock.	var/obj/item/paper/incident/active_report = null```
the same for the vars you’ve touched here

File: code/game/machinery/doors/brig_system.dm [hunk 2]
```diff
@@ -11,348 +17,333 @@
...
 
-	var/time = (I.incident.time_to_release - REALTIMEOFDAY)/10
-	if(!C.active_timer)
-		time = ((C.brig_sentence * 600) - C.time_served)/10
-	return max(0, time)
+	// Close cell doors.
+	for(var/obj/structure/machinery/door/window/brigdoor/door in targets)
+		if(door.density)
+			continue
+		INVOKE_ASYNC(door, /obj/structure/machinery/door.proc/close)
```
Comment (harryob) [inline code/game/machinery/doors/brig_system.dm:206]: TYPE_PROC_REF

File: code/game/machinery/doors/brig_system.dm [hunk 3]
```diff
@@ -11,348 +17,333 @@
...
 
-	usr.set_interaction(src)
 
-	switch(href_list["brig"])
-		if("incident")
-			current_report = locate(href_list["incident_report"])
-			current_menu = "incident_menu"
+/obj/structure/machinery/brig_cell/Initialize()
+	. = ..()
+	addtimer(CALLBACK(src, .proc/search_for_components), 20)
```
Comment (harryob) [inline code/game/machinery/doors/brig_system.dm:274]: PROC_REF

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/brig_system.dm [hunk 4]
```diff
@@ -9,350 +15,347 @@
...
-			targets += P
+	data["viewing_incident"] = viewed_report ? TRUE : FALSE
 
-	if(length(targets) == 0)
-		stat |= BROKEN
+	if (viewed_report)
+		var/datum/crime_incident/incident = viewed_report.incident
 
-	update_icon()
+		var/time_left = get_time_left(incident)
```
Comment (stanalbatross) [inline code/game/machinery/doors/brig_system.dm:59]: calculate in-UI to reduce proc overhead

File: code/game/machinery/doors/brig_system.dm [hunk 5]
```diff
@@ -9,350 +15,347 @@
...
 
-/obj/structure/machinery/brig_cell/process()
-	if(inoperable())
-		return
+		data["suspect"] = incident.criminal_name
+		data["progress"] = 1 - (time_left / 60) / incident.brig_sentence
```
Comment (stanalbatross) [inline code/game/machinery/doors/brig_system.dm:62]: do in-UI

File: code/game/machinery/doors/brig_system.dm [hunk 6]
```diff
@@ -9,350 +15,347 @@
...
+		data["sentence"] = (incident.status & BRIG_SENTENCE_PERMA) ? "Permabrig" : "[incident.brig_sentence] Minutes"
+		data["time_left"] = "[num2text((time_left / 60) % 60)]:[add_zero(num2text(time_left % 60), 2)]"
```
Comment (stanalbatross) [inline code/game/machinery/doors/brig_system.dm:64]: same

File: code/game/machinery/doors/brig_system.dm [hunk 7]
```diff
@@ -9,350 +15,347 @@
...
 
-	if(!incident_reports)
-		stop_processing()
+		// Flags.
+		data["active"]	= (incident.status & BRIG_SENTENCE_ACTIVE) // Is this timer currently ticking?
+		data["started"]	= (incident.time_served || data["active"]) // Has this timer been started at all?
```
Comment (stanalbatross) [inline code/game/machinery/doors/brig_system.dm:68]:
```suggestion
		data["started"]	= (incident.time_served || (incident.status & BRIG_SENTENCE_ACTIVE)) // Has this timer been started at all?
```
readability

File: code/game/machinery/doors/brig_system.dm [hunk 8]
```diff
@@ -9,350 +15,347 @@
...
+			data["status_class"] = "info"
+		else if (incident.status & BRIG_SENTENCE_PERMA)
+			data["status_text"] = "PERMABRIG"
+			data["status_class"] = "danger"
+		else if (incident.status & BRIG_SENTENCE_SERVED)
+			data["status_text"] = "SERVED"
+			data["status_class"] = "success"
+		else
+			data["status_text"] = null
+			data["status_class"] = null
```
Comment (stanalbatross) [inline code/game/machinery/doors/brig_system.dm:81]: why not just send the incident_status to the UI and calculate these all in-UI? look at the escape pod tgui for implementation example

File: code/game/machinery/doors/brig_system.dm [hunk 9]
```diff
@@ -9,350 +15,347 @@
...
-		ai_silent_announcement("BRIG REPORT: [C.criminal_name] has been jailed for [C.charges_to_string()].")
+// Pardon this incident.
+/obj/structure/machinery/brig_cell/proc/do_pardon(mob/user)
+	viewed_report.incident.status |= BRIG_SENTENCE_PARDONED
+	viewed_report.name += " (PARDONED)"
 
-/obj/structure/machinery/brig_cell/proc/timer_end(var/obj/item/paper/incident/I)
-	if(inoperable() || (!I && !current_report))
-		return FALSE
+	log_admin("[key_name(user)] has pardoned [viewed_report.incident.criminal_name] for [viewed_report.incident.charges_to_string()].")
```
Comment (stanalbatross) [inline code/game/machinery/doors/brig_system.dm:192]: JMP

File: code/game/machinery/doors/brig_system.dm [hunk 10]
```diff
@@ -9,350 +15,347 @@
...
+	var/datum/crime_incident/incident = active_report.incident
 
-/obj/structure/machinery/brig_cell/ui_interact(var/mob/user)
-	var/dat = "<HTML><BODY><TT>"
+	incident.status &= ~BRIG_SENTENCE_ACTIVE
+	incident.time_served = (incident.brig_sentence * 600) - (incident.time_to_release - REALTIMEOFDAY)
 
-	dat += create_controls()
-	dat += "<hr>"
+	log_admin("[key_name(user)] has paused the jail timer of [incident.criminal_name], [incident.charges_to_string()].")
```
Comment (stanalbatross) [inline code/game/machinery/doors/brig_system.dm:241]: JMP link should be included

File: code/game/machinery/doors/brig_system.dm [hunk 11]
```diff
@@ -9,350 +15,347 @@
...
 
+/obj/structure/machinery/brig_cell/attackby(obj/item/W, mob/living/user)
+	if(!istype(W, /obj/item/paper/incident))
+		return
+
+	user.temp_drop_inv_item(W)
+	W.moveToNullspace()
+	incident_reports += W
+
+	to_chat(user, SPAN_NOTICE("You insert [W] into [name]."))
```
Comment (stanalbatross) [inline code/game/machinery/doors/brig_system.dm:346]: \the [w] into \the [name]

File: tgui/packages/tgui/interfaces/BrigCell.js [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        <Button
+          icon="arrow-left"
+          px="2rem"
+          py=".25rem"
+          mr="1rem"
+          onClick={() => act('set_viewed_incident')}
+        />
+
+        <Flex.Item>
+          {status_text !== 'PERMABRIG' && (
```
Comment (stanalbatross) [inline tgui/packages/tgui/interfaces/BrigCell.js:93]: all these PERMABRIG checks could be done like `const Permabrig = status_text === 'PERMABRIG';`
would be a bit easier to read/write

**Review by stanalbatross — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/brig_system.dm [hunk 13]
```diff
@@ -9,350 +15,346 @@
...
+		if ("pause_timer")
+			timer_pause(usr)
+
+		if ("end_timer")
+			timer_end(usr)
+
+		if ("reset_timer")
+			timer_reset(usr)
+
+		if ("pardon")
```
Comment (harryob) [inline code/game/machinery/doors/brig_system.dm:148]: check for if they can pardon. i know the button appears dependent on this - but you can exploit tgui to send whatever ui_acts you like

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by harryob — APPROVED** [review]

**Pull request merged**